### PR TITLE
COMP: complete cargo features in `Cargo.toml`

### PR DESCRIPTION
**Pull request by vlad20012**
Continuation of #6204 - now with features completion in `Cargo.toml` !

Related to #4693

### CODE REVIEW

File: toml/src/main/kotlin/org/rust/toml/Util.kt [hunk 1]
```diff
@@ -92,6 +92,19 @@ class StringValueInsertionHandler(private val keyValue: TomlKeyValue) : InsertHa
     }
 }
 
+/** Wraps inserted string with quotes if needed */
+class StringLiteralInsertionHandler : InsertHandler<LookupElement> {
+    override fun handleInsert(context: InsertionContext, item: LookupElement) {
+        val leaf = context.getElementOfType<PsiElement>() ?: return
+        val hasQuotes = leaf.parent is TomlLiteral && leaf.elementType != TomlElementTypes.NUMBER
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/Util.kt:99]: Do I understand right that everything inside `TomlLiteral` is either string or number? Or what about other things like booleans? `hasQuotes` looks a bit confusing to me (maybe it works fine, but just looks a bit strange)

File: toml/src/main/kotlin/org/rust/toml/completion/CargoTomlFeatureDependencyCompletionProvider.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        val pkg = tomlFile.findCargoPackageForCargoToml() ?: return
+        for (dep in pkg.dependencies) {
+            if (dep.pkg.origin == PackageOrigin.STDLIB) continue
+            // TODO avoid AST loading?
+            for (feature in dep.pkg.getPackageTomlFile(tomlFile.project)?.allFeatures().orEmpty()) {
+                result.addElement(
+                    LookupElementBuilder
+                        .createWithSmartPointer("${dep.pkg.name}/${feature.text}", feature)
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/completion/CargoTomlFeatureDependencyCompletionProvider.kt:46]: Hmm, are you sure about using `dep.pkg.name` name here, not `dep.name`? Looks like this does not work for aliases like `aliased_dep = { package = "dep", version = "1.0" }`

File: toml/src/main/kotlin/org/rust/toml/completion/CargoTomlDependencyFeaturesCompletionProvider.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * features = ["<caret>"]
+ *             #^ Provides completion here
+ * ```
+ *
+ * @see [org.rust.toml.resolve.CargoTomlDependencyFeaturesReferenceProvider]
+ */
+class CargoTomlDependencyFeaturesCompletionProvider : CompletionProvider<CompletionParameters>() {
+    override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet) {
+        val parentParentParent = parameters.position.ancestorOrSelf<TomlArray>()?.parent?.parent as? TomlElement ?: return
+        val pkgName = if (parentParentParent is TomlTable) {
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/completion/CargoTomlDependencyFeaturesCompletionProvider.kt:37]: Code duplication (see `CargoTomlDependencyFeaturesReferenceProvider`)

**Review by ortem — CHANGES_REQUESTED** [review]

File: toml/src/main/kotlin/org/rust/toml/Util.kt (hunk 1 above)
Comment (vlad20012) [reply toml/src/main/kotlin/org/rust/toml/Util.kt:99]: Hmm, this was just copied from somewhere... Yes, I think it's better to write `val hasQuotes = leaf.parent is TomlLiteral && leaf.elementType in TOML_STRING_LITERALS` in this case

File: toml/src/main/kotlin/org/rust/toml/completion/CargoTomlFeatureDependencyCompletionProvider.kt (hunk 2 above)
Comment (vlad20012) [reply toml/src/main/kotlin/org/rust/toml/completion/CargoTomlFeatureDependencyCompletionProvider.kt]: Oh, it's bad. Looks like I made the same mistake in a lot of places :(

**Review by ortem — APPROVED** [review]

File: toml/src/main/kotlin/org/rust/toml/completion/CargoTomlFeatureDependencyCompletionProvider.kt (hunk 2 above)
Comment (vlad20012) [reply toml/src/main/kotlin/org/rust/toml/completion/CargoTomlFeatureDependencyCompletionProvider.kt:46]: It's not so simple even in this case. I'll fix this in another PR

**Comment by vlad20012** [comment]
bors r=ortem

**Pull request merged**